src/Features/Core/Portable/SolutionCrawler/WorkCoordinator.cs

```diff
@@ -421,6 +423,41 @@ private void EnqueueDocumentChangedEvent(Solution oldSolution, Solution newSolut
                         var newProject = newSolution.GetRequiredProject(documentId.ProjectId);
 
                         await EnqueueChangedDocumentWorkItemAsync(oldProject.GetRequiredDocument(documentId), newProject.GetRequiredDocument(documentId)).ConfigureAwait(false);
+
+                        // If all features are enabled for source generated documents, the solution crawler needs to
+                        // include them in incremental analysis.
+                        if (_syntaxTreeConfigurationService is { EnableOpeningSourceGeneratedFilesInWorkspace: true })
+                        {
+                            // TODO: if this becomes a hot spot, we should be able to expose/access the dictionary
+                            // underneath GetSourceGeneratedDocumentsAsync rather than create a new one here.
```

Comment:
I am not sure if all this code to detect added/removed/changed source generated documents belongs here. Shouldn't this already be retreivable from APIs on `ProjectChanges`? [https://github.com/dotnet/roslyn/blob/0098764330d1e22c43ead6065e4ed5f8aae09195/src/Workspaces/Core/Portable/Workspace/Solution/Project.cs#L470-L473](https://github.com/dotnet/roslyn/blob/0098764330d1e22c43ead6065e4ed5f8aae09195/src/Workspaces/Core/Portable/Workspace/Solution/Project.cs#L470-L473)

If not, then should be fix/update `ProjectChanges` to provide this data?

Reply:
➡️ ProjectChanges occurs prior to source generators running (and we don't want to delay it by an hour waiting for a slow generator).

Reply:
This might be a good request to at least add the diff APIs to ProjectChanges, like some sort of "GetAddedSourceGeneratedDocumentsAsync", but not sure we've had another need for that yet.